src/xma/src/xmaapi/xmahw_hal.cpp

```diff
@@ -123,159 +124,72 @@ bool hal_configure(XmaHwCfg *hwcfg, XmaXclbinParameter *devXclbins, int32_t num_
              return false;
          }    
 
-        uuid_copy(dev_tmp1.uuid, info.uuid); 
-        dev_tmp1.number_of_cus = info.number_of_kernels;
-        dev_tmp1.number_of_mem_banks = info.number_of_mem_banks;
-        if (dev_tmp1.number_of_cus > MAX_XILINX_KERNELS + MAX_XILINX_SOFT_KERNELS) {
-            xma_logmsg(XMA_ERROR_LOG, XMAAPI_MOD, "Could not download xclbin file %s to device %d\n",
-                        xclbin.c_str(), dev_index);
-            xma_logmsg(XMA_ERROR_LOG, XMAAPI_MOD, "XMA & XRT supports max of %d CUs but xclbin has %d number of CUs\n", MAX_XILINX_KERNELS + MAX_XILINX_SOFT_KERNELS, dev_tmp1.number_of_cus);
-            return false;
-        }
-        if (dev_tmp1.number_of_mem_banks > MAX_DDR_MAP) {
-            xma_logmsg(XMA_ERROR_LOG, XMAAPI_MOD, "XMA supports max of only %d mem banks\n", MAX_DDR_MAP);
-            return false;
-        }
-        dev_tmp1.kernels.reserve(dev_tmp1.number_of_cus);
-        dev_tmp1.ddrs.reserve(dev_tmp1.number_of_mem_banks);
-        dev_tmp1.number_of_hardware_kernels = info.number_of_hardware_kernels;
-
-        xma_logmsg(XMA_DEBUG_LOG, XMAAPI_MOD,"\nFor device id: %d; DDRs are:", dev_index);
-        auto& xma_mem_topology = info.mem_topology;
-        for (uint32_t d = 0; d < info.number_of_mem_banks; d++) {
-            dev_tmp1.ddrs.emplace_back(XmaHwMem{});
-            XmaHwMem& tmp1 = dev_tmp1.ddrs.back();
-            memset((void*)tmp1.name, 0x0, MAX_KERNEL_NAME);
-            xma_mem_topology[d].m_tag.copy((char*)tmp1.name, MAX_KERNEL_NAME-1);
-
-            tmp1.base_address = xma_mem_topology[d].m_base_address;
-            tmp1.size_kb = xma_mem_topology[d].m_size;
-            tmp1.size_mb = tmp1.size_kb / 1024;
-            tmp1.size_gb = tmp1.size_mb / 1024;
-            if (xma_mem_topology[d].m_used == 1 &&
-                tmp1.size_kb != 0 &&
-                (xma_mem_topology[d].m_type == MEM_TYPE::MEM_DDR3 || 
-                xma_mem_topology[d].m_type == MEM_TYPE::MEM_DDR4 ||
-                xma_mem_topology[d].m_type == MEM_TYPE::MEM_DRAM ||
-                xma_mem_topology[d].m_type == MEM_TYPE::MEM_HBM)
-                ) {
-                tmp1.in_use = true;
-                xma_logmsg(XMA_DEBUG_LOG, XMAAPI_MOD,"\tMEM# %d - %s - size: %lu KB", d, (char*)tmp1.name, tmp1.size_kb);
-            } else {
-                xma_logmsg(XMA_DEBUG_LOG, XMAAPI_MOD,"\tMEM# %d - %s - Unused/Unused Type", d, (char*)tmp1.name);
-
-            }
-        }
-
+        const axlf* xclbin_ax = reinterpret_cast<const axlf*>(buffer);      
+        std::vector<uint64_t> cu_addrs_sorted = xrt_core::xclbin::get_cus(info.ip_axlf, false);
+        uuid_copy(dev_tmp1.uuid, xclbin_ax->m_header.uuid);
         xma_logmsg(XMA_DEBUG_LOG, XMAAPI_MOD,"\nFor device id: %d; CUs are:", dev_index);
-        for (uint32_t d = 0; d < info.number_of_kernels; d++) {
+        for (int32_t d = 0; d < info.ip_axlf->m_count; d++) {
```

Comment:
Would it be possible to keep the axlf as an xrt::xclbin object and then use it for extracting the meta data?  It would be best if we keep all xclbin parsing consolidate to xrt::xclbin.
For example, xrt::xclbin allows iterations of all ips and cus, the latter being IP_KERNEL objects, so the loop could be replaced by a range-based-for if xrt::xclbin::get_cus ?

Reply:
Removed this logic and used XRT::XCLBIN class in the code.